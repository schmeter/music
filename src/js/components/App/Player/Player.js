import React from 'react';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';

import Analyser from './Analyser';

import { getNextIndex } from '../../../services/audio';

import { isTouch } from '../../../util/screen';

import configApp from '../../../../data/app.json';


class Player extends React.Component {
    audio = React.createRef();

    componentDidMount() {
        const { initAudio } = this.props;
        initAudio();
        this.loadAudio();
        document.addEventListener('keydown', this.captureKeys);
    }

    componentDidUpdate(lastProps) {
        const { activeIndex, initAudio, isLoggedIn, playToggle } = this.props;
        if (lastProps.isLoggedIn !== isLoggedIn) {
            initAudio();
        }
        if (lastProps.activeIndex !== activeIndex) {
            // accept values greater than -1, because -1 is initial index
            this.loadAudio(~lastProps.activeIndex);
        }
        if (lastProps.playToggle !== playToggle) {
            this.togglePlay();
        }
    }

    @autobind
    captureKeys(e) {
        switch (e.keyCode) {
            case 32:
                e.preventDefault();
                this.togglePlay();
                break;
            case 37:
                e.preventDefault();
                this.decreaseTimeFrame();
                break;
            case 39:
                e.preventDefault();
                this.increaseTimeFrame();
                break;
            default:
        }
    }

    loadAudio(play = false) {
        const { tracks, activeIndex, setActiveTrack } = this.props;
        const audio = this.audio.current;
        const file = tracks[activeIndex];
        if (file) {
            setActiveTrack(file);
            this.pause();
            audio.loop = !!file.loop;
            audio.src = file.path;
            audio.title = `${file.tag.artist} - ${file.tag.title}`;
            if (play) {
                this.play();
            } else {
                audio.load();
            }
        }
    }

    play() {
        const audio = this.audio.current;
        audio.play();
    }

    pause() {
        const audio = this.audio.current;
        audio.pause();
    }

    togglePlay() {
        const { isPlaying } = this.props;
        if (!isPlaying) {
            this.play();
        } else {
            this.pause();
        }
    }

    increaseTimeFrame() {
        const audio = this.audio.current;
        audio.currentTime += 5;
    }

    decreaseTimeFrame() {
        const audio = this.audio.current;
        audio.currentTime -= 5;
    }

    @autobind
    handlePlay() {
        const { setIsPlaying } = this.props;
        setIsPlaying(true);
    }

    @autobind
    handlePause() {
        const { setIsPlaying } = this.props;
        setIsPlaying(false);
    }

    @autobind
    handleError() {
        const { setIsPlaying } = this.props;
        setIsPlaying(false);
    }

    @autobind
    handleEnded() {
        const { activeIndex, tracks, setActiveIndex } = this.props;
        setActiveIndex(getNextIndex(activeIndex, tracks));
    }

    isAnalyserAllowed() {
        return !isTouch() && configApp.useAnalyser;
    }

    render() {
        const { isPlaying } = this.props;
        const audio = this.audio.current;
        const showAnalyser = this.isAnalyserAllowed() && audio;

        return (
            <div className="player">
                {!showAnalyser ? null : (
                    <Analyser
                        audio={audio}
                        isPlaying={isPlaying}
                    />
                )}
                <audio
                    className="audio"
                    preload="none"
                    controls
                    ref={this.audio}
                    onPlay={this.handlePlay}
                    onPause={this.handlePause}
                    onEnded={this.handleEnded}
                    onError={this.handleError}
                />
            </div>
        );
    }
}

Player.propTypes = {
    initAudio: PropTypes.func.isRequired,
    setActiveTrack: PropTypes.func.isRequired,
    setActiveIndex: PropTypes.func.isRequired,
    setIsPlaying: PropTypes.func.isRequired,
    tracks: PropTypes.array.isRequired,
    activeIndex: PropTypes.number.isRequired,
    playToggle: PropTypes.bool.isRequired,
    isPlaying: PropTypes.bool.isRequired,
    isLoggedIn: PropTypes.bool.isRequired
};

export default Player;
