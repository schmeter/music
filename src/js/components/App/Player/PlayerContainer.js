import { connect } from 'react-redux';

import Player from './Player';
import {
    initAudioAction,
    setActiveIndexAction,
    saveActiveTrackAction,
    setIsPlayingAction
} from '../../../state/actions/audio';

export const mapDispatchToProps = (dispatch) => ({
    initAudio: () => dispatch(initAudioAction()),
    setIsPlaying: (isPlaying) => dispatch(setIsPlayingAction(isPlaying)),
    setActiveTrack: (activeTrack) => dispatch(saveActiveTrackAction(activeTrack)),
    setActiveIndex: (activeIndex) => dispatch(setActiveIndexAction(activeIndex))
});

export const mapStateToProps = (state) => ({
    isLoggedIn: state.auth.isLoggedIn,
    tracks: state.audio.tracks,
    activeIndex: state.audio.activeIndex,
    playToggle: state.audio.playToggle,
    isPlaying: state.audio.isPlaying
});

export default connect(mapStateToProps, mapDispatchToProps)(Player);
