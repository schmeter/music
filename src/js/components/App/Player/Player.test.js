import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Player from './Player';

describe('Player', () => {
    window.HTMLMediaElement.prototype.load = jest.fn();
    window.HTMLMediaElement.prototype.play = jest.fn();
    window.HTMLMediaElement.prototype.pause = jest.fn();

    it('renders correctly without analyser', () => {
        const tracks = [{
            path: 'test1',
            tag: {
                title: 'test1',
            },
        }, {
            path: 'test2',
            tag: {
                title: 'test2',
            },
        }];
        const props = {
            config: {
                useAnalyser: false,
            },
            setActiveIndex: jest.fn(),
            setIsPlaying: jest.fn(),
            saveActiveTrack: jest.fn(),
            tracks,
            activeIndex: 0,
            nextIndex: 1,
            playToggle: false,
            isPlaying: false,
        };
        const component = mount(<Player {...props} />);

        expect(toJson(component)).toMatchSnapshot();

        component.setProps({
            ...props,
            activeIndex: 1,
            playToggle: false,
        });
        component.update();

        expect(toJson(component)).toMatchSnapshot();

        component.setProps({
            ...props,
            activeIndex: 1,
            playToggle: true,
        });
        component.update();

        expect(toJson(component)).toMatchSnapshot();

        component.setProps({
            ...props,
            tracks: [],
            activeIndex: 2,
            isPlaying: true,
        });
        component.update();

        expect(toJson(component)).toMatchSnapshot();

        document.dispatchEvent(new window.KeyboardEvent('keydown', { keyCode: 0 }));
        document.dispatchEvent(new window.KeyboardEvent('keydown', { keyCode: 32 }));

        component.setProps({
            ...props,
            isPlaying: false,
        });
        component.update();

        expect(toJson(component)).toMatchSnapshot();

        document.dispatchEvent(new window.KeyboardEvent('keydown', { keyCode: 32 }));
        component.find('audio').simulate('play');
        component.find('audio').simulate('pause');
        component.find('audio').simulate('ended');
        component.find('audio').simulate('error');
        document.dispatchEvent(new window.KeyboardEvent('keydown', { keyCode: 37 }));
        document.dispatchEvent(new window.KeyboardEvent('keydown', { keyCode: 39 }));
    });

    it('renders correctly with analyser', () => {
        const tracks = [{
            path: 'test1',
            tag: {
                title: 'test1',
            },
        }, {
            path: 'test2',
            tag: {
                title: 'test2',
            },
        }];
        const props = {
            config: {
                useAnalyser: true,
            },
            setActiveIndex: jest.fn(),
            setIsPlaying: jest.fn(),
            saveActiveTrack: jest.fn(),
            tracks,
            activeIndex: 0,
            nextIndex: 1,
            playToggle: false,
            isPlaying: false,
        };
        const component = mount(<Player {...props} />);

        expect(toJson(component)).toMatchSnapshot();
    });
});
