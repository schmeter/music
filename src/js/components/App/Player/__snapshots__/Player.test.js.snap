// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Player
  activeIndex={0}
  initAudio={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  isLoggedIn={false}
  isPlaying={false}
  playToggle={false}
  setActiveIndex={[MockFunction]}
  setActiveTrack={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "path": "test1",
            "tag": Object {
              "title": "test1",
            },
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  setIsPlaying={[MockFunction]}
  tracks={
    Array [
      Object {
        "path": "test1",
        "tag": Object {
          "title": "test1",
        },
      },
      Object {
        "path": "test2",
        "tag": Object {
          "title": "test2",
        },
      },
    ]
  }
>
  <div
    className="player"
  >
    <Analyser
      audio={null}
      isPlaying={false}
    />
    <audio
      className="audio"
      controls={true}
      onEnded={[Function]}
      onError={[Function]}
      onPause={[Function]}
      onPlay={[Function]}
      preload="none"
      tabIndex="-1"
    />
  </div>
</Player>
`;

exports[`renders correctly 2`] = `
<Player
  activeIndex={1}
  initAudio={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  isLoggedIn={false}
  isPlaying={false}
  playToggle={true}
  setActiveIndex={[MockFunction]}
  setActiveTrack={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "path": "test1",
            "tag": Object {
              "title": "test1",
            },
          },
        ],
        Array [
          Object {
            "path": "test2",
            "tag": Object {
              "title": "test2",
            },
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  setIsPlaying={[MockFunction]}
  tracks={
    Array [
      Object {
        "path": "test1",
        "tag": Object {
          "title": "test1",
        },
      },
      Object {
        "path": "test2",
        "tag": Object {
          "title": "test2",
        },
      },
    ]
  }
>
  <div
    className="player"
  >
    <Analyser
      audio={
        <audio
          class="audio"
          controls=""
          preload="none"
          src="test2"
          tabindex="-1"
          title="undefined - test2"
        />
      }
      isPlaying={false}
    >
      <canvas
        className="canvas"
      />
    </Analyser>
    <audio
      className="audio"
      controls={true}
      onEnded={[Function]}
      onError={[Function]}
      onPause={[Function]}
      onPlay={[Function]}
      preload="none"
      tabIndex="-1"
    />
  </div>
</Player>
`;

exports[`renders correctly 3`] = `
<Player
  activeIndex={2}
  initAudio={
    [MockFunction] {
      "calls": Array [
        Array [],
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  isLoggedIn={true}
  isPlaying={true}
  playToggle={true}
  setActiveIndex={[MockFunction]}
  setActiveTrack={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "path": "test1",
            "tag": Object {
              "title": "test1",
            },
          },
        ],
        Array [
          Object {
            "path": "test2",
            "tag": Object {
              "title": "test2",
            },
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  setIsPlaying={[MockFunction]}
  tracks={Array []}
>
  <div
    className="player"
  >
    <Analyser
      audio={
        <audio
          class="audio"
          controls=""
          preload="none"
          src="test2"
          tabindex="-1"
          title="undefined - test2"
        />
      }
      isPlaying={true}
    >
      <canvas
        className="canvas"
      />
    </Analyser>
    <audio
      className="audio"
      controls={true}
      onEnded={[Function]}
      onError={[Function]}
      onPause={[Function]}
      onPlay={[Function]}
      preload="none"
      tabIndex="-1"
    />
  </div>
</Player>
`;

exports[`renders correctly 4`] = `
<Player
  activeIndex={2}
  initAudio={
    [MockFunction] {
      "calls": Array [
        Array [],
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  isLoggedIn={true}
  isPlaying={false}
  playToggle={true}
  setActiveIndex={[MockFunction]}
  setActiveTrack={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "path": "test1",
            "tag": Object {
              "title": "test1",
            },
          },
        ],
        Array [
          Object {
            "path": "test2",
            "tag": Object {
              "title": "test2",
            },
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  setIsPlaying={[MockFunction]}
  tracks={Array []}
>
  <div
    className="player"
  >
    <Analyser
      audio={
        <audio
          class="audio"
          controls=""
          preload="none"
          src="test2"
          tabindex="-1"
          title="undefined - test2"
        />
      }
      isPlaying={false}
    >
      <canvas
        className="canvas"
      />
    </Analyser>
    <audio
      className="audio"
      controls={true}
      onEnded={[Function]}
      onError={[Function]}
      onPause={[Function]}
      onPlay={[Function]}
      preload="none"
      tabIndex="-1"
    />
  </div>
</Player>
`;
